File: 0278-first-bad-version/0278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int startIdx = 0;
        int endIdx = n;
        while (startIdx <= endIdx) {
            int midIdx = startIdx + (endIdx - startIdx) / 2;
            if (isBadVersion(midIdx)) {
                if (!isBadVersion(midIdx - 1)) {
                    return midIdx;
                }
                endIdx = midIdx - 1;
            } else {
                startIdx = midIdx + 1;
            } 
        }
        return 0;
    }
}

File: 0278-first-bad-version/0278-first-bad-version.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */

class Solution: VersionControl() {
    override fun firstBadVersion(n: Int) : Int {
        var startIdx = 0
        var endIdx = n
        while (startIdx <= endIdx) {
            val midIdx = startIdx + (endIdx - startIdx) / 2
            when {
                isBadVersion(midIdx) && !isBadVersion(midIdx - 1) -> return midIdx
                isBadVersion(midIdx) -> endIdx = midIdx - 1
                else -> startIdx = midIdx + 1
            }
        }
        return 0
	}
}